File: scripts/analysis/metaAnalysis.pl
#!/usr/bin/env perl
use strict;
use warnings;
use Cwd;
use lib $ENV{'GALACTICUS_ANALYSIS_PERL_PATH'}."/perl";
use PDL;
use PDL::NiceSlice;
use PDL::IO::HDF5;
use GnuPlot::LaTeX;
use GnuPlot::PrettyPlots;
use Data::Dumper;

# Create plots of collected metadata on merger tree ODE evolver.
# Andrew Benson (05-June-2011)

# Get arguments.
die ("Usage: metaAnalysis.pl <modelFile> <outputFolder>")
    unless ( scalar(@ARGV) == 2 );
my $modelFile    = $ARGV[0];
my $outputFolder = $ARGV[1];

# Open the file containing the meta-data.
my $HDFfile = new PDL::IO::HDF5($modelFile);

# Read the timestep histogram.
my $foundGroup      = 0;
my @groupsAvailable = $HDFfile->groups();
foreach my $groupAvailable ( @groupsAvailable ) {
    $foundGroup = 1 if ( $groupAvailable eq "metaData" );
}
die ("metaAnalysis.pl: metaData group does not exist") unless ( $foundGroup == 1 );
$foundGroup      = 0;
@groupsAvailable = $HDFfile->group("metaData")->groups();
foreach my $groupAvailable ( @groupsAvailable ) {
    $foundGroup = 1 if ( $groupAvailable eq "evolverProfiler" );
}
die ("metaAnalysis.pl: metaData/evolverProfile group does not exist") unless ( $foundGroup == 1 );
my $timeSteps                  = $HDFfile->group("metaData/evolverProfiler")->dataset("timeStep"                  )->get();
my $timeStepCount              = $HDFfile->group("metaData/evolverProfiler")->dataset("timeStepCount"             )->get();
my $evaluationCount            = $HDFfile->group("metaData/evolverProfiler")->dataset("evaluationCount"           )->get();
my $timeCPU                    = $HDFfile->group("metaData/evolverProfiler")->dataset("timeCPU"                   )->get();
my $timeStepCountInterrupted   = $HDFfile->group("metaData/evolverProfiler")->dataset("timeStepCountInterrupted"  )->get();
my $evaluationCountInterrupted = $HDFfile->group("metaData/evolverProfiler")->dataset("evaluationCountInterrupted")->get();
my $timeCPUInterrupted         = $HDFfile->group("metaData/evolverProfiler")->dataset("timeCPUInterrupted"        )->get();
my $propertyNames              = $HDFfile->group("metaData/evolverProfiler")->dataset("propertyNames"             )->get();
my $propertyHitCount           = $HDFfile->group("metaData/evolverProfiler")->dataset("propertyHitCount"          )->get();
my $propertyHitRate            = 100.0*$propertyHitCount/$propertyHitCount->sum();
my $propertyIndex              = pdl 1..nelem($propertyHitRate);

# Convert counts into frequencies.
my $timeStepFrequency                    = float(  $timeStepCount           )/sum(  $timeStepCount           );
my $timeStepWeightedFrequency            = float($evaluationCount           )/sum($evaluationCount           );
my $timeCPUFrequency                     = float(  $timeCPU                 )/sum(  $timeCPU                 );
my $timeStepInterruptedFrequency         = float(  $timeStepCountInterrupted)/sum(  $timeStepCountInterrupted);
my $timeStepInterruptedWeightedFrequency = float($evaluationCountInterrupted)/sum($evaluationCountInterrupted);
my $timeCPUInterruptedFrequency          = float(  $timeCPUInterrupted      )/sum(  $timeCPUInterrupted      );

# Find cumulative frequencies.
my $timeStepFrequencyCumulative                    = $timeStepFrequency                   ->cumusumover()/$timeStepFrequency                   ->sum();
my $timeStepWeightedFrequencyCumulative            = $timeStepWeightedFrequency           ->cumusumover()/$timeStepWeightedFrequency           ->sum();
my $timeCPUFrequencyCumulative                     = $timeCPUFrequency                    ->cumusumover()/$timeCPUFrequency                    ->sum();
my $timeStepInterruptedFrequencyCumulative         = $timeStepInterruptedFrequency        ->cumusumover()/$timeStepInterruptedFrequency        ->sum();
my $timeStepInterruptedWeightedFrequencyCumulative = $timeStepInterruptedWeightedFrequency->cumusumover()/$timeStepInterruptedWeightedFrequency->sum();
my $timeCPUInterruptedFrequencyCumulative          = $timeCPUInterruptedFrequency         ->cumusumover()/$timeCPUInterruptedFrequency         ->sum();

# Find ranges for timestep plot.
my $frequencies        = $timeStepFrequency->append($timeStepWeightedFrequency)->append($timeStepInterruptedFrequencyCumulative)->append($timeStepInterruptedWeightedFrequencyCumulative);
my $frequenciesNonZero = which($frequencies > 0.0);
my $timestepMinimum    = 10.0**(floor(log10(        $timeSteps  ->(( 0)               ) )));
my $timestepMaximum    = 10.0**(ceil (log10(        $timeSteps  ->((-1)               ) )));
my $frequencyMinimum   = 10.0**(floor(log10(minimum($frequencies->($frequenciesNonZero)))));
my $frequencyMaximum   = 10.0**(ceil (log10(maximum($frequencies->($frequenciesNonZero)))));

# Declare variables for GnuPlot;
my ($gnuPlot, $outputFile, $outputFileTeX, $plot);

# Open a pipe to GnuPlot and make a plot of timestep frequency.
$outputFileTeX = $outputFolder."/timesteps.tex";
open($gnuPlot,"|gnuplot 1>/dev/null 2>&1");
print $gnuPlot "set terminal cairolatex pdf standalone color lw 2 size 6in,4in\n";
print $gnuPlot "set output '".$outputFileTeX."'\n";
print $gnuPlot "set lmargin screen 0.15\n";
print $gnuPlot "set rmargin screen 0.95\n";
print $gnuPlot "set bmargin screen 0.15\n";
print $gnuPlot "set tmargin screen 0.95\n";
print $gnuPlot "set key spacing 1.2\n";
print $gnuPlot "set key at screen 0.2,0.2\n";
print $gnuPlot "set key left\n";
print $gnuPlot "set key bottom\n";
print $gnuPlot "set logscale xy\n";
print $gnuPlot "set mxtics 10\n";
print $gnuPlot "set format x '\$10^{\%L}\$'\n";
print $gnuPlot "set mytics 10\n";
print $gnuPlot "set format y '\$10^{\%L}\$'\n";
print $gnuPlot "set xrange [". $timestepMinimum.":". $timestepMaximum."]\n";
print $gnuPlot "set yrange [".$frequencyMinimum.":".$frequencyMaximum."]\n";
print $gnuPlot "set title offset 0,-0.8 'Timestep distribution'\n";
print $gnuPlot "set xlabel 'Timestep [Gyr]'\n";
print $gnuPlot "set ylabel 'Frequency'\n";
&GnuPlot::PrettyPlots::Prepare_Dataset
    (
     \$plot,
     $timeSteps,
     $timeStepFrequency,
     style  => "point",
     symbol => [6,7],
     weight => [3,1],
     title  => "Steps",
     color  => $GnuPlot::PrettyPlots::colorPairs{'cornflowerBlue'}
    );
&GnuPlot::PrettyPlots::Prepare_Dataset
    (
     \$plot,
     $timeSteps,
     $timeStepWeightedFrequency,
     style  => "point",
     symbol => [6,7],
     weight => [3,1],
     title  => "Evaluations",
     color  => $GnuPlot::PrettyPlots::colorPairs{'mediumSeaGreen'}
    );
&GnuPlot::PrettyPlots::Prepare_Dataset
    (
     \$plot,
     $timeSteps,
     $timeCPUFrequency,
     style  => "point",
     symbol => [6,7],
     weight => [3,1],
     title  => "CPU",
     color  => $GnuPlot::PrettyPlots::colorPairs{'indianRed'}
    );
&GnuPlot::PrettyPlots::Prepare_Dataset
    (
     \$plot,
     $timeSteps,
     $timeStepInterruptedFrequency,
     style  => "point",
     symbol => [6,6],
     weight => [3,1],
     title  => "Steps (interrupted)",
     color  => $GnuPlot::PrettyPlots::colorPairs{'cornflowerBlue'}
    );
&GnuPlot::PrettyPlots::Prepare_Dataset
    (
     \$plot,
     $timeSteps,
     $timeStepInterruptedWeightedFrequency,
     style  => "point",
     symbol => [6,6],
     weight => [3,1],
     title  => "Evaluations (interrupted)",
     color  => $GnuPlot::PrettyPlots::colorPairs{'mediumSeaGreen'}
    );
&GnuPlot::PrettyPlots::Prepare_Dataset
    (
     \$plot,
     $timeSteps,
     $timeStepInterruptedWeightedFrequency,
     style  => "point",
     symbol => [6,6],
     weight => [3,1],
     title  => "Evaluations (interrupted)",
     color  => $GnuPlot::PrettyPlots::colorPairs{'mediumSeaGreen'}
    );
&GnuPlot::PrettyPlots::Prepare_Dataset
    (
     \$plot,
     $timeSteps,
     $timeCPUInterruptedFrequency,
     style  => "point",
     symbol => [6,6],
     weight => [3,1],
     title  => "CPU (interrupted)",
     color  => $GnuPlot::PrettyPlots::colorPairs{'indianRed'}
    );
&GnuPlot::PrettyPlots::Plot_Datasets($gnuPlot,\$plot);
close($gnuPlot);
&GnuPlot::LaTeX::GnuPlot2PDF($outputFileTeX);

# Open a pipe to GnuPlot and make a plot of cumulative timestep frequency.
$outputFileTeX = $outputFolder."/timestepsCumulative.tex";
open($gnuPlot,"|gnuplot 1>/dev/null 2>&1");
print $gnuPlot "set terminal cairolatex pdf standalone color lw 2 size 6in,4in\n";
print $gnuPlot "set output '".$outputFileTeX."'\n";
print $gnuPlot "set lmargin screen 0.15\n";
print $gnuPlot "set rmargin screen 0.95\n";
print $gnuPlot "set bmargin screen 0.15\n";
print $gnuPlot "set tmargin screen 0.95\n";
print $gnuPlot "set key spacing 1.2\n";
print $gnuPlot "set key at screen 0.2,0.6\n";
print $gnuPlot "set key left\n";
print $gnuPlot "set key bottom\n";
print $gnuPlot "set logscale x\n";
print $gnuPlot "set mxtics 10\n";
print $gnuPlot "set format x '\$10^{\%L}\$'\n";
print $gnuPlot "set xrange [".$timestepMinimum.":".$timestepMaximum."]\n";
print $gnuPlot "set yrange [-0.02:1.02]\n";
print $gnuPlot "set title offset 0,-0.8 'Cumulative timestep distribution'\n";
print $gnuPlot "set xlabel 'Timestep [Gyr]'\n";
print $gnuPlot "set ylabel 'Cumulative fraction'\n";
&GnuPlot::PrettyPlots::Prepare_Dataset
    (
     \$plot,
     $timeSteps,
     $timeStepFrequencyCumulative,
     style  => "point",
     symbol => [6,7],
     weight => [3,1],
     title  => "Steps",
     color  => $GnuPlot::PrettyPlots::colorPairs{'cornflowerBlue'}
    );
&GnuPlot::PrettyPlots::Prepare_Dataset
    (
     \$plot,
     $timeSteps,
     $timeStepWeightedFrequencyCumulative,
     style  => "point",
     symbol => [6,7],
     weight => [3,1],
     title  => "Evaluations",
     color  => $GnuPlot::PrettyPlots::colorPairs{'mediumSeaGreen'}
    );
&GnuPlot::PrettyPlots::Prepare_Dataset
    (
     \$plot,
     $timeSteps,
     $timeCPUFrequencyCumulative,
     style  => "point",
     symbol => [6,7],
     weight => [3,1],
     title  => "CPU",
     color  => $GnuPlot::PrettyPlots::colorPairs{'indianRed'}
    );
&GnuPlot::PrettyPlots::Prepare_Dataset
    (
     \$plot,
     $timeSteps,
     $timeStepInterruptedFrequencyCumulative,
     style  => "point",
     symbol => [6,6],
     weight => [3,1],
     title  => "Steps (interrupted)",
     color  => $GnuPlot::PrettyPlots::colorPairs{'cornflowerBlue'}
    );
&GnuPlot::PrettyPlots::Prepare_Dataset
    (
     \$plot,
     $timeSteps,
     $timeStepInterruptedWeightedFrequencyCumulative,
     style  => "point",
     symbol => [6,6],
     weight => [3,1],
     title  => "Evaluations (interrupted)",
     color  => $GnuPlot::PrettyPlots::colorPairs{'mediumSeaGreen'}
    );
&GnuPlot::PrettyPlots::Prepare_Dataset
    (
     \$plot,
     $timeSteps,
     $timeCPUInterruptedFrequencyCumulative,
     style  => "point",
     symbol => [6,6],
     weight => [3,1],
     title  => "CPU (interrupted)",
     color  => $GnuPlot::PrettyPlots::colorPairs{'indianRed'}
    );
&GnuPlot::PrettyPlots::Plot_Datasets($gnuPlot,\$plot);
close($gnuPlot);
&GnuPlot::LaTeX::GnuPlot2PDF($outputFileTeX);

# Open a pipe to GnuPlot and make a plot of property hit rates.
$outputFileTeX = $outputFolder."/hitRates.tex";
open($gnuPlot,"|gnuplot 1>/dev/null 2>&1");
print $gnuPlot "set terminal cairolatex pdf standalone color lw 2 size 6in,4in\n";
print $gnuPlot "set output '".$outputFileTeX."'\n";
print $gnuPlot "set lmargin screen 0.15\n";
print $gnuPlot "set rmargin screen 0.95\n";
print $gnuPlot "set bmargin screen 0.15\n";
print $gnuPlot "set tmargin screen 0.95\n";
print $gnuPlot "set key spacing 1.2\n";
print $gnuPlot "set key at screen 0.2,0.2\n";
print $gnuPlot "set key left\n";
print $gnuPlot "set key bottom\n";
my $xExtent = nelem($propertyHitCount)+1;
print $gnuPlot "set xrange [0:".$xExtent."]\n";
my $yExtent = maximum($propertyHitRate)*1.05;
print $gnuPlot "set yrange [0:".$yExtent."]\n";
print $gnuPlot "set title offset 0,-0.8 'Hit Frequency'\n";
print $gnuPlot "set xlabel 'Property'\n";
print $gnuPlot "set ylabel 'Hit frequency [\\%]'\n";
&GnuPlot::PrettyPlots::Prepare_Dataset(\$plot,
			      $propertyIndex,$propertyHitRate,
			      style => "boxes", symbol => [6,7], weight => [3,1],
			      color => $GnuPlot::PrettyPlots::colorPairs{'cornflowerBlue'});
for (my $iProperty=0;$iProperty<nelem($propertyHitCount);++$iProperty) {
    print $gnuPlot "set label '".$propertyNames->atstr($iProperty)."' at ".$propertyIndex->index($iProperty).",0 front rotate by 90\n";
}
&GnuPlot::PrettyPlots::Plot_Datasets($gnuPlot,\$plot);
close($gnuPlot);
&GnuPlot::LaTeX::GnuPlot2PDF($outputFileTeX);

exit;

File: scripts/analysis/metaAnalysis.py
#!/usr/bin/env python3
import sys
import os
import numpy as np
import h5py
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt

# Create plots of collected metadata on merger tree ODE evolver.
# Andrew Benson (05-June-2011) [Perl]; ported to Python

if len(sys.argv) != 3:
    print("Usage: metaAnalysis.py <modelFile> <outputFolder>", file=sys.stderr)
    sys.exit(1)

modelFile    = sys.argv[1]
outputFolder = sys.argv[2]

# Create the directory.
os.mkdirs(outputFolder, exist_ok=True)

# Open the file containing the meta-data.
with h5py.File(modelFile, 'r') as HDFfile:
    if 'metaData' not in HDFfile:
        print("metaAnalysis.py: metaData group does not exist", file=sys.stderr)
        sys.exit(1)
    if 'evolverProfiler' not in HDFfile['metaData']:
        print("metaAnalysis.py: metaData/evolverProfiler group does not exist", file=sys.stderr)
        sys.exit(1)

    grp = HDFfile['metaData/evolverProfiler']

    timeSteps                  = grp['timeStep'                  ][()]
    timeStepCount              = grp['timeStepCount'             ][()]
    evaluationCount            = grp['evaluationCount'           ][()]
    timeCPU                    = grp['timeCPU'                   ][()]
    timeStepCountInterrupted   = grp['timeStepCountInterrupted'  ][()]
    evaluationCountInterrupted = grp['evaluationCountInterrupted'][()]
    timeCPUInterrupted         = grp['timeCPUInterrupted'        ][()]
    propertyNames_raw          = grp['propertyNames'             ][()]
    propertyHitCount           = grp['propertyHitCount'          ][()]

# Decode property names.
if propertyNames_raw.dtype.kind in ('S', 'O'):
    propertyNames = [n.decode() if isinstance(n, bytes) else str(n) for n in propertyNames_raw]
else:
    propertyNames = [str(n) for n in propertyNames_raw]

propertyHitRate  = 100.0 * propertyHitCount / propertyHitCount.sum()
propertyIndex    = np.arange(1, len(propertyHitRate) + 1)

# Convert counts into frequencies.
timeStepFrequency                    = timeStepCount.astype(float)            / timeStepCount.sum()
timeStepWeightedFrequency            = evaluationCount.astype(float)          / evaluationCount.sum()
timeCPUFrequency                     = timeCPU.astype(float)                  / timeCPU.sum()
timeStepInterruptedFrequency         = timeStepCountInterrupted.astype(float) / timeStepCountInterrupted.sum()
timeStepInterruptedWeightedFrequency = evaluationCountInterrupted.astype(float) / evaluationCountInterrupted.sum()
timeCPUInterruptedFrequency          = timeCPUInterrupted.astype(float)       / timeCPUInterrupted.sum()

# Find cumulative frequencies.
timeStepFrequencyCumulative                    = np.cumsum(timeStepFrequency)                    / timeStepFrequency.sum()
timeStepWeightedFrequencyCumulative            = np.cumsum(timeStepWeightedFrequency)            / timeStepWeightedFrequency.sum()
timeCPUFrequencyCumulative                     = np.cumsum(timeCPUFrequency)                     / timeCPUFrequency.sum()
timeStepInterruptedFrequencyCumulative         = np.cumsum(timeStepInterruptedFrequency)         / timeStepInterruptedFrequency.sum()
timeStepInterruptedWeightedFrequencyCumulative = np.cumsum(timeStepInterruptedWeightedFrequency) / timeStepInterruptedWeightedFrequency.sum()
timeCPUInterruptedFrequencyCumulative          = np.cumsum(timeCPUInterruptedFrequency)          / timeCPUInterruptedFrequency.sum()

# Find ranges for the timestep plot.
allFrequencies  = np.concatenate([timeStepFrequency, timeStepWeightedFrequency,
                                  timeCPUFrequency, timeStepInterruptedFrequency, 
                                  timeStepInterruptedWeightedFrequency, timeCPUInterruptedFrequency])
nonZero         = allFrequencies > 0.0
timestepMinimum = 10.0 ** np.floor(np.log10(timeSteps[ 0]))
timestepMaximum = 10.0 ** np.ceil (np.log10(timeSteps[-1]))
frequencyMinimum = 10.0 ** np.floor(np.log10(allFrequencies[nonZero].min()))
frequencyMaximum = 10.0 ** np.ceil (np.log10(allFrequencies[nonZero].max()))

cornflowerBlue  = '#6495ED'
mediumSeaGreen  = '#3CB371'
indianRed       = '#CD5C5C'

# --- Plot 1: Timestep frequency distribution ---
fig, ax = plt.subplots(figsize=(6, 4))
ax.plot(timeSteps, timeStepFrequency,                    'o', color=cornflowerBlue, markersize=4, label='Steps')
ax.plot(timeSteps, timeStepWeightedFrequency,            'o', color=mediumSeaGreen, markersize=4, label='Evaluations')
ax.plot(timeSteps, timeCPUFrequency,                     'o', color=indianRed,      markersize=4, label='CPU')
ax.plot(timeSteps, timeStepInterruptedFrequency,         's', color=cornflowerBlue, markersize=4, label='Steps (interrupted)')
ax.plot(timeSteps, timeStepInterruptedWeightedFrequency, 's', color=mediumSeaGreen, markersize=4, label='Evaluations (interrupted)')
ax.plot(timeSteps, timeCPUInterruptedFrequency,          's', color=indianRed,      markersize=4, label='CPU (interrupted)')
ax.set_xscale('log')
ax.set_yscale('log')
ax.set_xlim(timestepMinimum, timestepMaximum)
ax.set_ylim(frequencyMinimum, frequencyMaximum)
ax.set_title('Timestep distribution')
ax.set_xlabel('Timestep [Gyr]')
ax.set_ylabel('Frequency')
ax.legend(loc='lower left', fontsize=8)
plt.tight_layout()
plt.savefig(os.path.join(outputFolder, 'timesteps.pdf'))
plt.close()

# --- Plot 2: Cumulative timestep frequency distribution ---
fig, ax = plt.subplots(figsize=(6, 4))
ax.plot(timeSteps, timeStepFrequencyCumulative,                    'o', color=cornflowerBlue, markersize=4, label='Steps')
ax.plot(timeSteps, timeStepWeightedFrequencyCumulative,            'o', color=mediumSeaGreen, markersize=4, label='Evaluations')
ax.plot(timeSteps, timeCPUFrequencyCumulative,                     'o', color=indianRed,      markersize=4, label='CPU')
ax.plot(timeSteps, timeStepInterruptedFrequencyCumulative,         's', color=cornflowerBlue, markersize=4, label='Steps (interrupted)')
ax.plot(timeSteps, timeStepInterruptedWeightedFrequencyCumulative, 's', color=mediumSeaGreen, markersize=4, label='Evaluations (interrupted)')
ax.plot(timeSteps, timeCPUInterruptedFrequencyCumulative,          's', color=indianRed,      markersize=4, label='CPU (interrupted)')
ax.set_xscale('log')
ax.set_xlim(timestepMinimum, timestepMaximum)
ax.set_ylim(-0.02, 1.02)
ax.set_title('Cumulative timestep distribution')
ax.set_xlabel('Timestep [Gyr]')
ax.set_ylabel('Cumulative fraction')
ax.legend(loc='lower left', fontsize=8)
plt.tight_layout()
plt.savefig(os.path.join(outputFolder, 'timestepsCumulative.pdf'))
plt.close()

# --- Plot 3: Property hit rates ---
fig, ax = plt.subplots(figsize=(6, 4))
ax.bar(propertyIndex, propertyHitRate, color=cornflowerBlue)
for iProperty, name in enumerate(propertyNames):
    ax.text(propertyIndex[iProperty], 0, name, rotation=90, va='bottom', ha='center', fontsize=7)
ax.set_xlim(0, len(propertyHitCount) + 1)
ax.set_ylim(0, propertyHitRate.max() * 1.05)
ax.set_title('Hit Frequency')
ax.set_xlabel('Property')
ax.set_ylabel('Hit frequency [%]')
plt.tight_layout()
plt.savefig(os.path.join(outputFolder, 'hitRates.pdf'))
plt.close()
